### PR TITLE
Migrated db controller to API v2

### PR DESCRIPTION
**Pull request by naz**
refs #9866

### CODE REVIEW

File: core/server/api/shared/frame.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ class Frame {
             }
         }
 
+        this.options.rawInclude = apiConfig.rawInclude;
```
Comment (gargol) [inline core/server/api/shared/frame.js]: This field was added to allow preserving options in an array as it came in (same logic in v1 - https://github.com/TryGhost/Ghost/blob/7fafa1e152708a8b672bb6b19d9dacf5680aa2d3/core/server/api/v0.1/utils.js#L286):
```
                if (!convertOptions.forModel) {
                    options.include = utils.prepareInclude(options.include, allowedIncludes);
                } else {
                    options.withRelated = utils.prepareInclude(options.include, allowedIncludes);
                    delete options.include;
                }
```

Was trying to find a better place for this but ended up putting it into options for now. Definitelly would like to discuss better place to put it (or maybe way around it so we don't need it at all)

File: core/server/api/shared/headers.js [hunk 2]
```diff
@@ -31,6 +31,24 @@ const disposition = {
...
+     * parameter is governed by [RFC6266](http://tools.ietf.org/html/rfc6266#section-4.3).
+     *
+     * For encoding whitespace and non-ISO-8859-1 characters, you MUST use the "filename*=" attribute, NOT "filename=".
+     * Ideally, both. Examples: http://tools.ietf.org/html/rfc6266#section-5
+     *
+     * We'll use ISO-8859-1 characters here to keep it simple.
+     *
+     * @see http://tools.ietf.org/html/rfc598
+     */
+    file(result) {
```
Comment (gargol) [inline core/server/api/shared/headers.js]: Needed for https://github.com/TryGhost/Ghost/pull/10051/files#diff-5aec6ddaa47f7b2024e989d0cf8af73fR44

File: core/server/api/shared/serializers/input/all.js [hunk 3]
```diff
@@ -21,8 +21,12 @@ module.exports = {
         debug('serialize all');
 
         if (frame.options.include) {
-            frame.options.withRelated = trimAndLowerCase(frame.options.include);
-            delete frame.options.include;
+            if (frame.options.rawInclude) {
```
Comment (gargol) [inline core/server/api/shared/serializers/input/all.js]: Logic come from https://github.com/TryGhost/Ghost/blob/7fafa1e152708a8b672bb6b19d9dacf5680aa2d3/core/server/api/v0.1/utils.js#L286

File: core/server/api/v2/utils/serializers/output/db.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            ? response[1].problems
+            : response[0].problems;
+
+        frame.response = {
+            db: [],
+            problems: problems
+        };
+    },
+
+    deleteAllContent(response, apiConfig, frame) {
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/output/db.js]: Even though `deleteAllContent` responds with `statusCode: 204` this transform is needed in case the method is used internally.

File: core/server/web/api/v2/admin/routes.js [hunk 5]
```diff
@@ -195,8 +199,6 @@ module.exports = function apiRoutes() {
         apiv2.http(apiv2.upload.image)
     );
 
-    router.post('/db/backup', mw.authenticateClient('Ghost Backup'), api.http(api.db.backupContent));
```
Comment (gargol) [inline core/server/web/api/v2/admin/routes.js]: Moved this route up to the rest of `/db` endpoints so they are grouped nicely

File: core/server/api/shared/frame.js [hunk 6]
```diff
@@ -37,6 +37,10 @@ class Frame {
             }
         }
 
+        if (apiConfig.rawInclude) {
```
Comment (gargol) [inline core/server/api/shared/frame.js]: https://github.com/TryGhost/Ghost/pull/10051#discussion_r227787385 still applies to this part of code

File: core/server/api/v2/utils/serializers/output/db.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        frame.response = {
+            db: [{filename: filename}]
+        };
+    },
+
+    exportContent([exportedData, fileName], apiConfig, frame) {
+        debug('exportContent');
+
+        frame.response = {
+            filename: fileName,
```
Comment (gargol) [inline core/server/api/v2/utils/serializers/output/db.js]: Was added here to decouple logic in http headers from services. Now it's up to the controller to return a filename

Comment (gargol) [inline core/server/api/v2/utils/serializers/output/db.js]: Need to rethink how to pass filename differently, so it's not included in the response object but still available for headers stage to be used :thinking:

File: core/server/api/shared/http.js [hunk 8]
```diff
@@ -21,7 +21,8 @@ const http = (apiImpl) => {
 
         frame.configure({
             options: apiImpl.options,
-            data: apiImpl.data
+            data: apiImpl.data,
+            rawInclude: apiImpl.rawInclude
```
Comment (kirrg001) [inline core/server/api/shared/http.js]: have you thought about a skip strategy for shared serializers?

Comment (gargol) [reply core/server/api/shared/http.js]: Do you mean conditionally skipping this step in serialization process - https://github.com/TryGhost/Ghost/blob/master/core/server/api/shared/serializers/handle.js#L28 ?

Not sure I fully understand how that would be done within the frame :thinking: ? Can you please expand?

Comment (gargol) [reply core/server/api/shared/http.js]: moved this logic to `handle.js` and now conditionally skipping input serialization logic based on `frame.options.rawInput` flag -> 987eb057a99181c05dec17f2f43d0c7703070e9e

File: core/server/api/shared/serializers/input/all.js [hunk 9]
```diff
@@ -1,17 +1,8 @@
 const debug = require('ghost-ignition').debug('api:shared:serializers:input:all');
```
Comment (gargol) [inline core/server/api/shared/serializers/input/all.js]: This changeset could be extracted and merged as a separate PR. But for now leaving it here for discussion and context

File: core/server/api/shared/headers.js [hunk 10]
```diff
@@ -37,15 +37,33 @@ const disposition = {
...
+     * For encoding whitespace and non-ISO-8859-1 characters, you MUST use the "filename*=" attribute, NOT "filename=".
+     * Ideally, both. Examples: http://tools.ietf.org/html/rfc6266#section-5
+     *
+     * We'll use ISO-8859-1 characters here to keep it simple.
+     *
+     * @see http://tools.ietf.org/html/rfc598
+     */
+    file(result, options, frame = {}) {
+        return {
+            'Content-Disposition': `Attachment; filename="${frame.file}"`
```
Comment (gargol) [inline core/server/api/shared/headers.js]: Had tried multiple approaches and this one seems like the most appropriate in this situation. The main reason frame is used for filename storage is that we don't want to modify the api and include filename in `result` object. The second reason is that the operation of filename calculation is asynchronous and it would need a lot of reshuffling to do so (see `Object.assign(headers, disposition[apiConfig.disposition.type](result, apiConfig.disposition, frame));`). The second reason also eliminates calculating the filename directly here like csv handler does above - `if (typeof options.value === 'function') {...`

**Comment by gargol** [comment]
@kirrg001 would be great if you could have a look at this one :+1:  let's try and shrink this list https://github.com/TryGhost/Ghost/issues/10060 one bite at a time :smiley:

**Comment by kirrg001** [comment]
@gargol Will do today 👍

File: core/server/web/api/v2/admin/routes.js (hunk 5 above)
Comment (kirrg001) [reply core/server/web/api/v2/admin/routes.js]: 👍

File: core/server/api/shared/frame.js [hunk 11]
```diff
@@ -37,6 +37,10 @@ class Frame {
             }
         }
 
+        if (apiConfig.rawInput) {
+            this.options.rawInput = apiConfig.rawInput;
```
Comment (kirrg001) [inline core/server/api/shared/frame.js]: Even if the shared serializer would transform `include` to `withRelated`, you always have access to `frame.original.include` or? 🤔

File: core/server/api/shared/headers.js (hunk 10 above)
Comment (kirrg001) [reply core/server/api/shared/headers.js]: Thanks for sharing 👍 Okay i see the problem. I am just concerned that we now use two different approaches to get header values 🤔 

Let's go back. Why is the exporter filename async? Because it reads the title from the database.
But the settings is always cached in progress. (see settings cache)

I see two other solutions:
1. the exporter uses `settingsCache.get('title')`, which makes the call sync and keeps consistenty with the current header value concept. Could be a short term solution, but not a full future proof solution.

2. As you said, make the header generation async. As soon as we have another async case we will run into this problem again. So this is a future proof solution.

---

I see two problems with attaching a value to the frame:

1. It's very hard to read. We just attach a random value `frame.file`. What does this mean? This would need a proper grouping e.g. `frame.header.value`.

2. If we want to treat header generation as a completely separate "pipe" later on, then we tie controller logic and header logic together. (headers === http, controller logic === functional/http)

---

I am fine with keeping as is + raise an issue to refactor. But i have the feeling we are raising more and more refactoring issues and we don't have much time working on them 😆

Open for any other ideas/solutions whatever 😎

File: core/server/api/shared/serializers/input/all.js [hunk 12]
```diff
@@ -21,17 +12,17 @@ module.exports = {
         debug('serialize all');
 
         if (frame.options.include) {
-            frame.options.withRelated = trimAndLowerCase(frame.options.include);
+            frame.options.withRelated = utils.options.trimAndLowerCase(frame.options.include);
```
[the remark this answers is not in the document]
Comment (gargol) [reply core/server/api/shared/serializers/input/all.js]: Opened a PR for this change here https://github.com/TryGhost/Ghost/pull/10263 :)

File: core/server/api/v2/db.js [hunk 13]
```diff
@@ -23,6 +22,7 @@ module.exports = {
             }
         },
         query(frame) {
+            Object.assign(frame.options, {include: frame.options.withRelated});
```
Comment (kirrg001) [inline core/server/api/v2/db.js]: That is much better 👍 , even though it's hard to understand. I know it sucks that we have two uses cases for `include`, but from user perspective it makes sense i think.

Another option is to make serializers configureable in the controller e.g.

```
serializers: {
  input: {
    all: {
       toModelOptions: false
    }
  }
}
```
You can also access `frame.original.query.include`, BUT that value is not validated/serialized ☹️

Comment (gargol) [reply core/server/api/v2/db.js]: > BUT that value is not validated/serialized
Yeah, that's exactly why didn't go that way :+1:

Comment (kirrg001) [reply core/server/api/v2/db.js]: Let's just use this approach for now 👍 Maybe add a comment to the code base?

Comment (gargol) [reply core/server/api/v2/db.js]: :+1: comment added :)

File: core/server/api/shared/headers.js (hunk 10 above)
Comment (gargol) [reply core/server/api/shared/headers.js]: I'm trying the approach of making header assignment async. The filename generation will be similar to what `csv` header does - a call to async `options.file()` that will be defined on the controller. If that becomes too complicated will try doing the same approach but with sync `settingsCache.get('title')` call.

Comment (gargol) [reply core/server/api/shared/headers.js]: The other idea would be having some extra property defined on a `frame` that would serve as a solution to keep data that dosn't really belong to the response.. but from experience those places usually become kind of a 'dump' of all sorts of things, so would rather avoid it :laughing:

File: core/server/api/shared/http.js [hunk 14]
```diff
@@ -27,6 +27,10 @@ const http = (apiImpl) => {
 
         apiImpl(frame)
             .then((result) => {
+                return shared.headers.get(result, apiImpl.headers, frame)
+                    .then(headers => ({result, headers}));
```
Comment (gargol) [inline core/server/api/shared/http.js]: This is an ugly bit imo. @kirrg001 maybe you have a better idea on a more elegant syntax for passing multiple objects further into `.then` chain?

**Comment by gargol** [comment]
@kirrg001 think, main issues were addressed, so this PR is ready for a final review :crossed_fingers:

File: core/server/api/shared/headers.js [hunk 15]
```diff
@@ -37,22 +37,75 @@ const disposition = {
...
 };
 
 module.exports = {
     get(result, apiConfig = {}) {
         let headers = {};
 
-        if (apiConfig.disposition) {
-            Object.assign(headers, disposition[apiConfig.disposition.type](result, apiConfig.disposition));
-        }
+        return Promise.resolve()
```
Comment (gargol) [inline core/server/api/shared/headers.js]: This set of changes could go into master separately as 'Async support for headers' . Lmk if should create another PR for it :)

Comment (kirrg001) [reply core/server/api/shared/headers.js]: I'd suggest to extract the header async refactoring to master straight? But maybe next Monday? Need to start with QA for tomorrow's release :)

Comment (gargol) [reply core/server/api/shared/headers.js]: sounds good :+1: I'm happy with current state of this PR as is, finally see the end of this over month-long battle to migrate a single controller :laughing:

Comment (kirrg001) [reply core/server/api/shared/headers.js]: 🤫🤣

**Comment by kirrg001** [comment]
@gargol What is the state of this PR? I lost track what's left.

**Comment by gargol** [comment]
The header handling for files has to be extracted into separate PR (doing it at the moment) and then a final review of this PR containing a bit fewer changes. I think it's ready to be merged but let's have another look after I break things up

**Comment by gargol** [comment]
@kirrg001 it's finally nice and clean 1 commit PR :tada: Let me know if this is mergeable :dancing_women:

**Review by kirrg001 — APPROVED** [review]
Just left one comment, otherwise feel free to self merge 👍

**Comment by gargol** [comment]
My Christmas wish: "to kill this PR before the end of 2018" came true :smile:
